File: windows/src/HoverPocket.Shell/Capabilities/CapabilityApprovalPresentation.cs
using System.Globalization;
using System.Text;
using HoverPocket.Shell.Providers.Sticky;

namespace HoverPocket.Shell.Capabilities;

internal enum CapabilityApprovalTargetState
{
    Present,
    Missing
}

// Host-memory-only. Do not serialize this into contracts, audit logs, receipts,
// generated surfaces, or agent transcripts.
internal sealed record CapabilityApprovalPresentation(
    string RequestId,
    string PlanDigest,
    string StepId,
    string ArgumentDigest,
    string ActionKey,
    string TargetKind,
    string TargetDisplayKey,
    string? TargetDisplayLabel,
    CapabilityApprovalTargetState TargetState,
    bool Destructive,
    bool RollbackAvailable);

internal interface ICapabilityApprovalPresentationResolver
{
    IReadOnlyList<CapabilityApprovalPresentation> Resolve(
        CapabilityExecutionPlan plan,
        IReadOnlyList<PocketCapabilityDescriptor> descriptors,
        CapabilityApprovalRequest request);
}

internal sealed class EmptyCapabilityApprovalPresentationResolver : ICapabilityApprovalPresentationResolver
{
    public IReadOnlyList<CapabilityApprovalPresentation> Resolve(
        CapabilityExecutionPlan plan,
        IReadOnlyList<PocketCapabilityDescriptor> descriptors,
        CapabilityApprovalRequest request) => [];
}

internal sealed class HostCapabilityApprovalPresentationResolver(StickyNotesStore stickyStore)
    : ICapabilityApprovalPresentationResolver
{
    private const int MaximumDisplayRunes = 80;

    public IReadOnlyList<CapabilityApprovalPresentation> Resolve(
        CapabilityExecutionPlan plan,
        IReadOnlyList<PocketCapabilityDescriptor> descriptors,
        CapabilityApprovalRequest request)
    {
        var presentations = new List<CapabilityApprovalPresentation>();
        foreach (var pair in plan.Steps.Zip(descriptors))
        {
            if (pair.Second.Key != CapabilityIds.StickyDelete)
            {
                continue;
            }
            var rawId = CapabilityJson.RequiredString(pair.First.Arguments, "noteId", 128);
            var effect = request.Effects.SingleOrDefault(item => item.StepId == pair.First.Id);
            if (!Guid.TryParse(rawId, out var noteId) || effect is null)
            {
                throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "approval_target");
            }
            var note = stickyStore.GetNote(noteId);
            presentations.Add(new CapabilityApprovalPresentation(
                request.Id,
                request.PlanDigest,
                pair.First.Id,
                effect.ArgumentDigest,
                "approval.sticky.note.delete",
                "sticky_note",
                note is null ? "approval.target.sticky_note.missing" : "approval.target.sticky_note",
                note is null ? null : SanitizedDisplayLabel(note.Title),
                note is null ? CapabilityApprovalTargetState.Missing : CapabilityApprovalTargetState.Present,
                pair.Second.Effect == CapabilityEffect.DestructiveSensitive,
                pair.Second.RollbackAvailable));
        }
        return presentations;
    }

    internal static string? SanitizedDisplayLabel(string value)
    {
        var result = new StringBuilder();
        var runeCount = 0;
        var pendingSpace = false;
        foreach (var rune in value.EnumerateRunes())
        {
            if (Rune.IsWhiteSpace(rune))
            {
                pendingSpace = result.Length > 0;
                continue;
            }
            var category = Rune.GetUnicodeCategory(rune);
            if (category is UnicodeCategory.Control or UnicodeCategory.Format)
            {
                continue;
            }
            if (pendingSpace && runeCount < MaximumDisplayRunes)
            {
                result.Append(' ');
                runeCount++;
                pendingSpace = false;
            }
            if (runeCount >= MaximumDisplayRunes)
            {
                break;
            }
            result.Append(rune.ToString());
            runeCount++;
        }
        var sanitized = result.ToString().Trim();
        return sanitized.Length == 0 ? null : sanitized;
    }
}

File: Sources/HoverPocket/Capabilities/CapabilityApprovalPresentation.swift
import Foundation

enum CapabilityApprovalTargetState: String, Equatable, Sendable {
    case present
    case missing
}

/// Host-memory-only presentation data. It must not be encoded into contracts,
/// audit logs, receipts, generated surfaces, or agent transcripts.
struct CapabilityApprovalPresentation: Equatable, Sendable {
    let requestID: String
    let planDigest: String
    let stepID: String
    let argumentDigest: String
    let actionKey: String
    let targetKind: String
    let targetDisplayKey: String
    let targetDisplayLabel: String?
    let targetState: CapabilityApprovalTargetState
    let destructive: Bool
    let rollbackAvailable: Bool
}

@MainActor
protocol CapabilityApprovalPresentationResolving {
    func resolve(
        plan: CapabilityExecutionPlan,
        descriptors: [PocketCapabilityDescriptor],
        request: CapabilityApprovalRequest
    ) throws -> [CapabilityApprovalPresentation]
}

@MainActor
struct EmptyCapabilityApprovalPresentationResolver: CapabilityApprovalPresentationResolving {
    func resolve(
        plan _: CapabilityExecutionPlan,
        descriptors _: [PocketCapabilityDescriptor],
        request _: CapabilityApprovalRequest
    ) throws -> [CapabilityApprovalPresentation] {
        []
    }
}

@MainActor
final class HostCapabilityApprovalPresentationResolver: CapabilityApprovalPresentationResolving {
    private static let maximumDisplayScalars = 80
    private let stickyStore: StickyNotesStore

    init(stickyStore: StickyNotesStore) {
        self.stickyStore = stickyStore
    }

    func resolve(
        plan: CapabilityExecutionPlan,
        descriptors: [PocketCapabilityDescriptor],
        request: CapabilityApprovalRequest
    ) throws -> [CapabilityApprovalPresentation] {
        var presentations: [CapabilityApprovalPresentation] = []
        for (step, descriptor) in zip(plan.steps, descriptors) where descriptor.key == PocketCapabilityKeys.stickyDelete {
            let rawID = try step.arguments.requiredString("noteId", maxLength: 128)
            guard let noteID = UUID(uuidString: rawID),
                  let effect = request.effects.first(where: { $0.stepID == step.id }) else {
                throw CapabilityBrokerError.invalidPlan("approval_target")
            }
            let note = stickyStore.note(id: noteID)
            presentations.append(CapabilityApprovalPresentation(
                requestID: request.id,
                planDigest: request.planDigest,
                stepID: step.id,
                argumentDigest: effect.argumentDigest,
                actionKey: "approval.sticky.note.delete",
                targetKind: "sticky_note",
                targetDisplayKey: note == nil
                    ? "approval.target.sticky_note.missing"
                    : "approval.target.sticky_note",
                targetDisplayLabel: note.flatMap { Self.sanitizedDisplayLabel($0.title) },
                targetState: note == nil ? .missing : .present,
                destructive: descriptor.effect == .destructiveSensitive,
                rollbackAvailable: descriptor.rollbackAvailable
            ))
        }
        return presentations
    }

    static func sanitizedDisplayLabel(_ value: String) -> String? {
        var result = ""
        var scalarCount = 0
        var pendingSpace = false
        for scalar in value.unicodeScalars {
            if CharacterSet.whitespacesAndNewlines.contains(scalar) {
                pendingSpace = !result.isEmpty
                continue
            }
            if CharacterSet.controlCharacters.contains(scalar)
                || scalar.properties.generalCategory == .format {
                continue
            }
            if pendingSpace, scalarCount < maximumDisplayScalars {
                result.append(" ")
                scalarCount += 1
                pendingSpace = false
            }
            guard scalarCount < maximumDisplayScalars else { break }
            result.unicodeScalars.append(scalar)
            scalarCount += 1
        }
        let sanitized = result.trimmingCharacters(in: .whitespacesAndNewlines)
        return sanitized.isEmpty ? nil : sanitized
    }
}
